phantom_token.c

```diff
@@ -488,202 +502,96 @@ static ngx_int_t handler(ngx_http_request_t *request)
-}
-
 static ngx_int_t introspection_response_handler(ngx_http_request_t *request, void *data,
                                                 ngx_int_t introspection_subrequest_status_code)
 {
     phantom_token_module_context_t *module_context = (phantom_token_module_context_t*)data;
+    ngx_str_t cache_data = ngx_null_string;
+    u_char *jwt_start = NULL;
+    size_t jwt_len = 0;
+    size_t bearer_jwt_len = 0;
+    u_char *p = NULL;
+    bool use_buffer_response = false;
 
-    ngx_log_debug1(NGX_LOG_DEBUG_HTTP, request->connection->log, 0, "auth request done status = %d",
+    ngx_log_error(NGX_LOG_DEBUG, request->connection->log, 0, "Introspection request done status = %d",
                    request->headers_out.status);
 
     module_context->status = request->headers_out.status;
 
-    // fail early for not 200 response
+    // Fail early if the introspection request returned a non 200 response
     if (request->headers_out.status != NGX_HTTP_OK)
     {
         module_context->done = 1;
-
         return introspection_subrequest_status_code;
     }
 
-    u_char *jwt_start = NULL;
-    ngx_str_t cache_data = ngx_null_string;
-
 #if (NGX_HTTP_CACHE)
     if (request->cache && !request->cache->buf)
     {
         // We have a cache but it's not primed
         ngx_http_file_cache_open(request);
     }
 
-    if (jwt_start == NULL && request->cache && request->cache->buf && request->cache->valid_sec > 0)
+    // When caching is enabled the JWT is always received from the cache, including the initial request from a client with a new opaque access token
+    if (request->cache && request->cache->buf && request->cache->valid_sec > 0)
     {
-        // Try to read JWT from cache
-
         cache_data.len = request->cache->length;
         cache_data.data = ngx_pnalloc(request->pool, cache_data.len);
 
-        if (cache_data.data != NULL)
+        if (cache_data.data == NULL)
+        {
+            ngx_log_error(NGX_LOG_ERR, request->connection->log, 0, "Failed to allocate memory for cache data");
+        }
+        else
         {
             ngx_read_file(&request->cache->file, cache_data.data, cache_data.len, request->cache->body_start);
-
             jwt_start = cache_data.data;
+            jwt_len = request->headers_out.content_length_n;
         }
     }
     else
     {
-        jwt_start = request->header_end + sizeof("\r\n") - 1; // FIXME: Won't work if JWT is large
+        use_buffer_response = true;
     }
+#else
+    use_buffer_response = true;
+#endif
 
-    if (jwt_start == NULL)
+    // When caching is disabled the JWT is always read from upstream buffers
+    if (use_buffer_response)
     {
-        ngx_log_debug0(NGX_LOG_DEBUG_HTTP, request->connection->log, 0,
-                       "Failed to obtain JWT from introspection response or, if applicable, cache");
+        if (request->upstream->buffer.last == request->upstream->buffer.end)
```

Comment:
This logic was provided by the customer and seems correct. However, I would like to find a link that shows we are doing this in the correct standard way. This will also help use to do code reviews of this pull request.